### PR TITLE
[Simplified Collect][Workflows] Select workspace approver

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Fixed Issues

$ https://github.com/Expensify/App/issues/37195
PROPOSAL:

### Verify that you can set approver
1. Go to workspace settings
2. Select a control policy workspace that you're an admin of
3. Enable add approvals 
4. Click approver
5. Select a user
6. Verify that the approver is changed and persisted.
7. Open the approver screen again, verify that the approver selected in step 5 has a green check next to it.

### Verify that filtering (searching) works
1. Go to approver page
2. Search for a member, verify that the members are filtered.
3. Verify that you can select a searched member and it's set as an approver

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

image

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

image
</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/29673073/022302c5-3cf6-4704-a81c-074443bf07a7

image

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by rushatgabhane** [comment]
Lint and typecheck unrelated to PR. Please ignore them and proceed with review.

**Comment by allroundexperts** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by allroundexperts** [comment]
**BUG**
`No result found` error not displaying in the member search bar

Screenshot 2024-02-28 at 4 02 25 PM

**Comment by allroundexperts** [comment]
**BUG**
`No result found` shows up without entering anything inside the search bar?

Screenshot 2024-02-28 at 4 04 45 PM

**Comment by allroundexperts** [comment]
**BUG**

Reloading the approver selection page changes the page in the background.

https://github.com/Expensify/App/assets/30054992/a9d6304f-9bb1-4690-a714-d5f0047a0b2b

**Comment by allroundexperts** [comment]
**BUG**
`Admin` or `Owner` tag displaying at incorrect position.

Screenshot 2024-02-28 at 4 08 53 PM

Screenshot 2024-02-28 at 4 09 38 PM

**Comment by allroundexperts** [comment]
**BUG**

Approver email being shown instead of name.

Screenshot 2024-02-28 at 4 10 03 PM

**Comment by rushatgabhane** [comment]
> Admin or Owner tag displaying at incorrect position

expected. the mockup is a bit outdated

> Approver email being shown instead of name

not part of issue?

> Reloading the approver selection page changes the page in the background

@allroundexperts i need help with this. What might be causing this? Maybe `lastViewedPage` isn't updated? How to debug this?

**Comment by rushatgabhane** [comment]
@allroundexperts all done!

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx [hunk 1]
```diff
@@ -10,6 +10,7 @@ import useNetwork from '@hooks/useNetwork';
 import useStyleUtils from '@hooks/useStyleUtils';
 import useThemeStyles from '@hooks/useThemeStyles';
 import useWindowDimensions from '@hooks/useWindowDimensions';
+import Navigation from '@libs/Navigation/Navigation';
```
Comment (luacmartins) [inline src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:13]: Lint is still failing on this file

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx [hunk 2]
```diff
@@ -32,8 +34,7 @@ function WorkspaceWorkflowsPage({policy, route}: WorkspaceWorkflowsPageProps) {
     const {isSmallScreenWidth} = useWindowDimensions();
     const {isOffline} = useNetwork();
 
-    const ownerPersonalDetails = ReportUtils.getDisplayNamesWithTooltips(OptionsListUtils.getPersonalDetailsForAccountIDs([policy?.ownerAccountID ?? 0], CONST.EMPTY_OBJECT), false);
-    const policyOwnerDisplayName = ownerPersonalDetails[0]?.displayName;
+    const policyApprover = policy?.approver;
```
Comment (luacmartins) [inline src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:37]: I think we still wanna keep the default as `ownerPersonalDetails` if there's no explicit approver set in the policy

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+        Object.entries(policyMembers ?? {}).forEach(([accountIDKey, policyMember]) => {
+            const accountID = Number(accountIDKey);
+            if (isDeletedPolicyMember(policyMember)) {
+                return;
+            }
+
+            const details = personalDetails?.[accountID];
+            if (!details) {
+                Log.hmmm(`[WorkspaceMembersPage] no personal details found for policy member with accountID: ${accountID}`);
```
Comment (luacmartins) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:66]: I don't think we need to log anything here, we can just do an early return

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    const StyleUtils = useStyleUtils();
+
+    const isDeletedPolicyMember = (policyMember: PolicyMember): boolean =>
+        !isOffline && policyMember.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE && isEmptyObject(policyMember.errors);
+
+    const [formattedPolicyMembers, formattedApprover] = useMemo(() => {
+        const policyMemberDetails: MemberOption[] = [];
+        const approverDetails: MemberOption[] = [];
+
+        Object.entries(policyMembers ?? {}).forEach(([accountIDKey, policyMember]) => {
```
Comment (luacmartins) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:58]: This seems to be a copy/pasta from https://github.com/Expensify/App/blob/59f8f3d4889afe26c3f73c2e8e3df1ced50e11dd/src/pages/workspace/WorkspaceMembersPage.tsx#L282. Can we just extract that into a method and reuse it in both places?

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            };
+
+            if (policy?.approver === details.login) {
+                approverDetails.push(formattedMember);
+            } else {
+                policyMemberDetails.push(formattedMember);
+            }
+        });
+        return [policyMemberDetails, approverDetails];
+    }, [personalDetails, policyMembers, translate, policy?.approver]);
```
Comment (luacmartins) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:111]: missing dependencies

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    const sections: MembersSection[] = useMemo(() => {
+        const sectionsArr: MembersSection[] = [];
```
Comment (luacmartins) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:114]:
```suggestion
        const sectionsArray: MembersSection[] = [];
```

**Review by luacmartins — CHANGES_REQUESTED** [review]
Left a few comments. 

As for changing `policy.approver` to store an accountID instead of email, I don't think that's possible now since that's how the policy is stored in our database. We should probably create a method to get the details from personalDetails by email instead of accountID.

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx (hunk 2 above)
Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:37]: The default approver is still policy's owner.
You can test this by disabling approvals, and then re-enable them.

**Comment by rushatgabhane** [comment]
> We should probably create a method to get the details from personalDetails by email instead of accountID

@luacmartins this method will be kinda expensive. It's going to search for a email in the entire `personalDetails` object.

Would this be okay?

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx (hunk 1 above)
Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:13]: It's failing on `main` too. I'm hoping it'll be fixed soon on `main` and we can merge it

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx (hunk 3 above)
Comment (luacmartins) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:66]: I see we do this in other pages too, so let's leave it as is.

**Comment by luacmartins** [comment]
@rushatgabhane let me know when you've addressed my comments above.

**Comment by allroundexperts** [comment]
**BUG**
1. `Offline` indicator not at the bottom of the page
2. The selection is greyed out when offline even if the value was persisted in the db.

Screenshot 2024-03-01 at 12 04 22 AM

**Comment by allroundexperts** [comment]
Seems pretty easily reproducible. 

1. Open the workspace workflows page.
2. Go offline.

https://github.com/Expensify/App/assets/30054992/3f8a1a08-2ec5-49b1-aeb6-7aef5f0e7e50

**Comment by rushatgabhane** [comment]
@allroundexperts i think you might have some pending updates. The issue isn't reproducible with steps you've mentioned

https://github.com/Expensify/App/assets/29673073/09757bc5-b11b-4a6e-8699-f83d736d621f

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx (hunk 4 above)
Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:58]: @luacmartins i agree that this method shares most of the code. But I don't think it's a good idea to create a method that would reuse common logic.

1. Both methods in the page are unrelated and DRYing would couple them together. So it'll be harder to change them later
2. The two methods are different in the following ways: 
     1. approverpage: pushes the formatted option to 2 arrays. members page pushes to 1 array. 
     2. The disabled logic for both pages is different. 


This is the only common code - I'm not sure what to name this or how would this be reused in other places.
What do you think?

```js

const accountID = Number(accountIDKey);
if (isDeletedPolicyMember(policyMember)) {
    return;
}

const details = personalDetails?.[accountID];

if (!details) {
    Log.hmmm(`[WorkspaceMembersPage] no personal details found for policy member with accountID: ${accountID}`);
    return;
}

if (PolicyUtils.isExpensifyTeam(details?.login ?? details?.displayName)) {
    if (policyOwner && currentUserLogin && !PolicyUtils.isExpensifyTeam(policyOwner) && !PolicyUtils.isExpensifyTeam(currentUserLogin)) {
        return;
    }
}

const isOwner = policy?.owner === details.login;
const isAdmin = session?.email === details.login || policyMember.role === CONST.POLICY.ROLE.ADMIN;

let roleBadge = null;
if (isOwner || isAdmin) {
    roleBadge = (
        <Badge
            text={isOwner ? translate('common.owner') : translate('common.admin')}
            textStyles={styles.textStrong}
            badgeStyles={[styles.justifyContentCenter, StyleUtils.getMinimumWidth(60), styles.badgeBordered]}
        />
    );
}
```

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+            title: translate('common.all'),
+            data: formattedPolicyMembers,
+            shouldShow: true,
+            indexOffset: formattedApprover.length,
+        });
+
+        return sectionsArray;
+    }, [formattedPolicyMembers, formattedApprover, searchTerm, translate]);
+
+    const headerMessage = useMemo(() => (searchTerm && !sections[0].data.length ? translate('common.noResultsFound') : ''), [translate, sections]);
```
Comment (rushatgabhane) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:150]: Not a lint error. We don't want to recompute header every time a key is pressed

**Comment by rushatgabhane** [comment]
@luacmartins updated to show display name

image

**Comment by allroundexperts** [comment]
> @allroundexperts i need help with this. What might be causing this? How to debug this? I tried comparing the implementation of invite members page but couldn't find a difference.

I have spent some good time on it but wasn't able to figure out the reason myself 😄. One thing which I found out was that this does not work for other pages as well:

https://github.com/Expensify/App/assets/30054992/77999f2f-c54c-42eb-ba1f-ac0282881933

**Comment by rushatgabhane** [comment]
@allroundexperts thankss it's a broader issue. 
Issue to fix it all here - https://github.com/Expensify/App/issues/37517

**Comment by allroundexperts** [comment]
@rushatgabhane Can a member be an approver?

**Comment by rushatgabhane** [comment]
@allroundexperts yes

**Comment by luacmartins** [comment]
>@rushatgabhane Can a member be an approver?

yes, they can

**Comment by rushatgabhane** [comment]
Do you all see 89 commits in this PR? Github is going bonkers for me
#### My branch - 89 commits ahead
image


#### PR - 86 commits ahead
image

**Comment by luacmartins** [comment]
I see 86 commits

**Comment by rushatgabhane** [comment]
@luacmartins this would be a sync issue on github then. My branch is `89` commits ahead https://github.com/rushatgabhane/exfy/tree/approval-page

**Comment by luacmartins** [comment]
AH I was wondering what's up with that. I see a lot of the actions not being triggered

**Comment by allroundexperts** [comment]
Does this checkbox look right to you guys? The bottom of it seems to be stripped.

Screenshot 2024-03-01 at 8 00 59 PM

**Comment by rushatgabhane** [comment]
@allroundexperts not a part of this PR. It looks intentional to me

File: src/libs/PersonalDetailsUtils.ts [hunk 8]
```diff
@@ -56,6 +56,10 @@ function getPersonalDetailsByIDs(accountIDs: number[], currentUserAccountID: num
     return result;
 }
 
+function getPersonalDetailByEmail(email: string): PersonalDetails | undefined {
+    return Object.values(allPersonalDetails ?? {}).find((detail) => detail?.login === email) as unknown as PersonalDetails;
```
Comment (allroundexperts) [inline src/libs/PersonalDetailsUtils.ts:60]: Doesn't the following work?
```
(Object.values(allPersonalDetails ?? {}) as PersonalDetails[]).find((detail) => detail?.login === email)
```

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx [hunk 9]
```diff
@@ -96,8 +96,8 @@ function WorkspaceWorkflowsPage({policy, betas, route}: WorkspaceWorkflowsPagePr
                         title={translate('workflowsPage.approver')}
                         titleStyle={styles.textLabelSupportingNormal}
                         descriptionTextStyle={styles.textNormalThemeText}
-                        description={policyOwnerDisplayName ?? ''}
-                        // onPress={() => Navigation.navigate(ROUTES.WORKSPACE_WORKFLOWS_APPROVER.getRoute(route.params.policyID))}
+                        description={policyApproverName ?? ''}
+                        onPress={() => Navigation.navigate(ROUTES.WORKSPACE_WORKFLOWS_APPROVER.getRoute(route.params.policyID))}
                         // TODO will be done in https://github.com/Expensify/Expensify/issues/368334
```
Comment (allroundexperts) [inline src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:101]: Is this `TODO` still valid?

Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx:101]: nice catch! removed

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+    const setPolicyApprover = (member: MemberOption) => {
+        if (!policy || !policy.approvalMode || !personalDetails?.[member.accountID]?.login) {
```
Comment (allroundexperts) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:153]: `if (!policy?.approvalMode ||`

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+            return;
+        }
+        const approver: string = personalDetails?.[member.accountID]?.login ?? policy.approver ?? policy.owner;
```
Comment (allroundexperts) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:156]: Hm... Do we want this fallback? IMO, if `personalDetails?.[member.accountID]` is empty, we should just not let the user submit.

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    const sections: MembersSection[] = useMemo(() => {
+        const sectionsArray: MembersSection[] = [];
+
+        if (searchTerm !== '') {
+            const filteredOptions = [...formattedApprover, ...formattedPolicyMembers].filter((option) => {
+                const parsedPhoneNumber = parsePhoneNumber(LoginUtils.appendCountryCode(Str.removeSMSDomain(searchTerm)));
+                const searchValue = parsedPhoneNumber.possible ? parsedPhoneNumber.number?.e164 ?? '' : searchTerm.toLowerCase();
+
+                const isPartOfSearchTerm = !!option.text?.toLowerCase().includes(searchValue) || !!option.login?.toLowerCase().includes(searchValue);
+                return isPartOfSearchTerm;
```
Comment (allroundexperts) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:122]: I feel like this is being used at a lot of places. Is there a way we can make this a utility function?

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx (hunk 11 above)
Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:156]: On main, when approver toggle is enabled, policy owner is set as the default approver. So I don't think we should change that

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx (hunk 12 above)
Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:122]: I don't think we should.

The above snippet is different from the below one.  I don't  see the value in DRYing this

https://github.com/Expensify/App/blob/511a5ffc93e9afea65070e727d7df59093f10a40/src/pages/workspace/WorkspaceInvitePage.tsx#L177-L183

Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx]: What should it be called?

Comment (allroundexperts) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx]: How does `filterMembersBySearchTerm` sound?

**Comment by allroundexperts** [comment]
@rushatgabhane You can run lint and type checks locally as well before pushing the code. Not only will it be faster but it will also save us the CI minutes!

File: src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx (hunk 9 above)
Comment (luacmartins) [reply src/pages/workspace/workflows/WorkspaceWorkflowsPage.tsx]: yes, we should remove it

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx (hunk 12 above)
Comment (rushatgabhane) [reply src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx]: done

**Comment by rushatgabhane** [comment]
> You can run lint and type checks locally

: ( will ask help on slack
image

**Comment by allroundexperts** [comment]
Have you tried cleaning everything and installing again?

**Comment by rushatgabhane** [comment]
@allroundexperts all typechecks and lint errors fixed. I didn't fix errors that are also present on `main` branch.

image

**Comment by allroundexperts** [comment]
@rushatgabhane All checks are passing on main.

**Comment by rushatgabhane** [comment]
@allroundexperts i didn't make any changes in `withPolicy`

**Comment by allroundexperts** [comment]
> @allroundexperts we didn't make any changes in `withPolicy`. But there's still an error

Hm... Something must have changed. Because this is passing on main.

**Comment by allroundexperts** [comment]
TS passed 🎉

**Review by allroundexperts — APPROVED** [review]
Looks good. Great work @rushatgabhane!

**Review by luacmartins — APPROVED** [review]
LGTM

**Comment by luacmartins** [comment]
Created internal issue to send approverAccountID so we can get rid of [getPersonalDetailByEmail](https://github.com/Expensify/App/pull/37391/files#diff-ff1a0c0087f151243c758ae7346cbf8f25b22de7c61370102de747a06bb94cedR59) - https://github.com/Expensify/Expensify/issues/374749

File: src/libs/Navigation/types.ts [hunk 13]
```diff
@@ -63,6 +63,9 @@ type CentralPaneNavigatorParamList = {
     [SCREENS.WORKSPACE.WORKFLOWS]: {
         policyID: string;
     };
+    [SCREENS.WORKSPACE.WORKFLOWS_APPROVER]: {
+        policyID: string;
+    };
```
[the remark this answers is not in the document]
Comment (rushatgabhane) [reply src/libs/Navigation/types.ts:68]: could you please link the TS check

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+        return [policyMemberDetails, approverDetails];
+    }, [personalDetails, policyMembers, translate, policy?.approver, StyleUtils, isDeletedPolicyMember, policy?.owner, styles]);
+
+    const sections: MembersSection[] = useMemo(() => {
+        const sectionsArray: MembersSection[] = [];
+
+        if (searchTerm !== '') {
+            const filteredOptions = [...formattedApprover, ...formattedPolicyMembers].filter((option) => {
+                const searchValue = OptionsListUtils.getSearchValueForPhoneOrEmail(searchTerm);
+                return !!option.text?.toLowerCase().includes(searchValue) || !!option.login?.toLowerCase().includes(searchValue);
```
Comment (eh2077) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:119]: This line caused a regression here [Custom name user searched with email id shows no results found](https://github.com/Expensify/App/issues/42576#top)

File: src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+            return;
+        }
+        const approver: string = personalDetails?.[member.accountID]?.login ?? policy.approver ?? policy.owner;
+        Policy.setWorkspaceApprovalMode(policy.id, approver, policy.approvalMode);
+        Navigation.goBack();
+    };
+
+    return (
+        <ScreenWrapper
+            includeSafeAreaPaddingBottom={false}
```
Comment (c3024) [inline src/pages/workspace/workflows/WorkspaceWorkflowsApproverPage.tsx:165]: This causes extra space in Android if we pass this as `false` for a page with a `SelectionList` child containing a footer.

**Pull request merged**